google/cloud/storage/testing/retry_tests.h

```diff
@@ -39,6 +43,35 @@ ::testing::Matcher<Status> StoppedOnTooManyTransients(char const* api_name);
 /// Validates the `Status` produced in a "permanent error" test.
 ::testing::Matcher<Status> StoppedOnPermanentError(char const* api_name);
 
+/// Validates the idempotency tokens used in a retry loop.
+::testing::Matcher<std::vector<std::string>> RetryLoopUsesSingleToken();
+
+/// Captures the idempotency token. Refactors some code in
+/// MockRetryClientFunction.
+void CaptureIdempotencyToken(std::vector<std::string>& tokens,
+                             rest_internal::RestContext const& context);
+
+/// Captures information to validate the RetryClient loops and return a
+/// transient error.
+class MockRetryClientFunction {
+ public:
+  explicit MockRetryClientFunction(Status status);
+
+  std::vector<std::string> const& captured_tokens() const { return *tokens_; }
+
+  template <typename Request>
+  Status operator()(rest_internal::RestContext& context, Options const&,
+                    Request const&) {
+    CaptureIdempotencyToken(*tokens_, context);
+    return status_;
+  }
+
+ private:
+  std::shared_ptr<std::vector<std::string>> tokens_ =
+      std::make_shared<std::vector<std::string>>();
```

Comment:
The `shared_ptr`-ness doesn't seem to escape this class, so why not a value?

Reply:
Because (AFAICT) these are copied by `WillRepeatedly()` and `WillOnce()`.  We want them to share the captured tokens to examine them after the mocking / test took place.

Reply:
Hmm.  Then perhaps some form of "must be shared between copied instances" comment would be appropriate.

Reply:
Done.